### PR TITLE
Cranelift: Implement tail calls for x86_64

### PR DESCRIPTION
**Pull request by fitzgen**

### CODE REVIEW

**Comment by afonso360** [comment]
👋 Hey, I'm not sure if this is ready or not, but I implemented return call's on fuzzgen (#6641) and it found something.

It is also possible that this is related to #6640, but it feels quite different. Here we get a segfault, and we don't need the stack probe.

<details>
<summary>Testcase</summary>

```
test run
set preserve_frame_pointers=true
target x86_64

function %b(i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8) -> i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8 tail {
    ss0 = explicit_slot 126
    ss1 = explicit_slot 126
    ss2 = explicit_slot 13

block0(v0: i8, v1: i8, v2: i8, v3: i8, v4: i8, v5: i8, v6: i8, v7: i8, v8: i8, v9: i8, v10: i8, v11: i8, v12: i8, v13: i8, v14: i8, v15: i8):
    return v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v9, v9, v9, v9, v9, v9
}


function %a(i8 uext, i8 uext, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8) -> i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8 tail {
    sig0 = (i8 uext, i8 uext, i8 uext, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8) -> i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8 tail
    fn0 = %b sig0

block0(v0: i8, v1: i8, v2: i8, v3: i8, v4: i8, v5: i8, v6: i8, v7: i8, v8: i8, v9: i8, v10: i8, v11: i8, v12: i8, v13: i8, v14: i8, v15: i8):
    return_call fn0(v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0, v0)
}

; run: %a(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0) == [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```
</details>

**Comment by fitzgen** [comment]
@afonso360 still haven't dug in fully yet, but it is interesting that the second function claims the first function's signature is something other than what it the function is declared as: the second function says the first three arguments are `uext` but the first function doesn't declare that. That shouldn't matter here I don't think but seems like something that fuzzgen shouldn't generate.

**Comment by afonso360** [comment]
Oh boy, I removed `uext` and `sext` when I was minimizing the test case, but it looks like I accidentally left those in. Please ignore them! Fuzzgen should be generating those correctly, at least I haven't ever caught it not doing so.

I've updated the example above to remove them, and verified that it still reproduces the original issue.

**Comment by fitzgen** [comment]
Fixed that fuzz bug, thanks again for reporting it @afonso360!

Turns out that my codegen-the-arguments sequence was accidentally skipping "hidden" extra parameters like return pointers. Fixed!

File: cranelift/codegen/src/isa/x64/inst/emit.rs [hunk 1]
```diff
@@ -1584,6 +1585,55 @@ pub(crate) fn emit(
...
+                state,
+                call_info.new_stack_arg_size,
+                call_info.old_stack_arg_size,
+                call_info.ret_addr,
+                call_info.fp,
+                call_info.tmp,
+                &call_info.uses,
+            );
+
+            // Finally, jump to the callee!
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/emit.rs:1605]: A comment here noting why not delegate to `Inst::JmpKnown` (in the usual pseudoinst way, i.e. `let inst = ...; inst.emit(...);`) would be good. I think the main reason is that we have different metadata in this case, in particular we don't have a label for the target but rather a different kind of relocation; is that right?

File: cranelift/codegen/src/isa/x64/inst/emit.rs [hunk 2]
```diff
@@ -4089,3 +4139,156 @@ pub(crate) fn emit(
...
+    //            | ...                 |
+    //            | old stack slots     |
+    //            | ...                 |
+    //            +---------------------+
+    //            | ...                 |
+    //    new     | new stack arguments |
+    //    frame   | ...                 | <-- SP
+    //            +---------------------+
+    //
+    // We need to pop FP, copy the new stack arguments over the old
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/emit.rs:4208]: possibly just a semantic quibble but "pop FP" confused me here: to pop means to adjust SP, but here we're restoring FP's old value without adjusting the stack at all. Could we write it that way ("restore FP") instead?

File: cranelift/codegen/src/isa/x64/inst/emit.rs [hunk 3]
```diff
@@ -4089,3 +4139,156 @@ pub(crate) fn emit(
...
+        }
+        .emit(&[], sink, info, state);
+        Inst::MovRM {
+            size: OperandSize::Size64,
+            src: tmp,
+            dst: SyntheticAmode::Real(Amode::ImmReg {
+                // Add 2 because we need to skip over the old FP and the
+                // return address.
+                simm32: callee_sp_relative_to_fp.wrapping_add((i + 2) * 8),
+                base: fp,
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/emit.rs:4259]: I'm possibly confused here, but: isn't this using the wrong FP, i.e., the value we already restored above, to whatever the original caller had in RBP when we were called? I think we want to use *our* FP as a frame of reference here.

File: cranelift/codegen/src/machinst/buffer.rs [hunk 4]
```diff
@@ -406,7 +406,7 @@ impl<I: VCodeInst> MachBuffer<I> {
 
     /// Add a byte.
     pub fn put1(&mut self, value: u8) {
-        trace!("MachBuffer: put byte @ {}: {:x}", self.cur_offset(), value);
+        // trace!("MachBuffer: put byte @ {}: {:x}", self.cur_offset(), value);
```
Comment (cfallin) [inline cranelift/codegen/src/machinst/buffer.rs:409]: Commented-out lines here -- we should restore these or else delete them if we decide we don't need them (here and below).

File: cranelift/codegen/src/isa/x64/inst/emit.rs (hunk 3 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:4259]: Ah, actually, after reading further, I see that this is `fp` which is set up in the lowering helper to be a copy (`MovFromPReg`) of `rbp`, so this makes sense.

**Review by cfallin — APPROVED** [review]
Overall this looks totally reasonable to me -- excited to see the final piece fall into place, at least for x86-64!

In addition to comments above I think it would be nice to see the "extra ret-addr movement" go away for the overwhelmingly common case (no stack args) either in this PR or in an immediate followup -- it seems to me that it shouldn't be too much work (tens of lines of code? `Option<Reg>` for `ret_addr`, and check if old_size == new_size?) and it's an important (in)efficiency to clean up soon when using tail-calls as a sort of fine-grained control flow mechanism (BB jumps between funclets etc).

Otherwise, happy to see this merged with comments addressed!

File: cranelift/codegen/src/isa/x64/inst/emit.rs (hunk 1 above)
Comment (fitzgen) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:1605]: That's correct. Added a comment.

File: cranelift/codegen/src/isa/x64/inst/emit.rs (hunk 2 above)
Comment (fitzgen) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:4208]: Done.

File: cranelift/codegen/src/isa/x64/inst/emit.rs (hunk 3 above)
Comment (fitzgen) [reply cranelift/codegen/src/isa/x64/inst/emit.rs:4259]: (Oh I just hadn't refreshed the tab. Doh.)

**Comment by fitzgen** [comment]
> In addition to comments above I think it would be nice to see the "extra ret-addr movement" go away for the overwhelmingly common case (no stack args) either in this PR or in an immediate followup -- it seems to me that it shouldn't be too much work (tens of lines of code? `Option<Reg>` for `ret_addr`, and check if old_size == new_size?) and it's an important (in)efficiency to clean up soon when using tail-calls as a sort of fine-grained control flow mechanism (BB jumps between funclets etc).

Will do this in an immediate follow up.

Thanks for the review!

**Pull request merged**